Objects/obmalloc.c

```diff
@@ -1533,31 +1535,65 @@ _PyMem_DebugRawRealloc(void *ctx, void *p, size_t nbytes)
     }
     total = nbytes + 4*SST;
 
+    tail = q + original_nbytes;
+    serialno = read_size_t(tail + SST);
+    /* Mark the header, the trailer, ERASED_SIZE bytes at the begin and
+       ERASED_SIZE bytes at the end as dead and save the copy of erased bytes.
+     */
+    if (original_nbytes <= sizeof(save)) {
+        memcpy(save, q, original_nbytes);
+        memset(q - 2*SST, DEADBYTE, original_nbytes + 4*SST);
+    }
+    else {
+        memcpy(save, q, ERASED_SIZE);
+        memset(q - 2*SST, DEADBYTE, ERASED_SIZE + 2*SST);
+        memcpy(&save[ERASED_SIZE], tail - ERASED_SIZE, ERASED_SIZE);
+        memset(tail - ERASED_SIZE, DEADBYTE, ERASED_SIZE + 2*SST);
+    }
+
     /* Resize and add decorations. */
-    q = (uint8_t *)api->alloc.realloc(api->alloc.ctx, q - 2*SST, total);
-    if (q == NULL) {
-        return NULL;
+    r = (uint8_t *)api->alloc.realloc(api->alloc.ctx, q - 2*SST, total);
+    if (r == NULL) {
+        nbytes = original_nbytes;
```

Comment:
Please restore erased bytes here, and exit here.

For example, memcpy(q, save, Py_MIN(nbytes, original_nbytes)) below is wrong if realloc() failed: you must restore "original_nbytes" bytes, not Py_MIN(nbytes, original_nbytes) bytes.

Reply:
This is why I have added `nbytes = original_nbytes` here. `memcpy(q, save, Py_MIN(nbytes, original_nbytes)) below is correct, it restores "original_nbytes" bytes.

Reply:
Oh ok. But I still dislike all the code between realloc failure and code to restore erased byte executed even if realloc failed: 

```
    write_size_t(q, nbytes);
    q[SST] = (uint8_t)api->api_id;
    memset(q + SST + 1, FORBIDDENBYTE, SST-1);
    q += 2*SST;

    tail = q + nbytes;
    memset(tail, FORBIDDENBYTE, SST);
    write_size_t(tail + SST, serialno);
```

realloc() function is not supposed to modify anything if it fails.

Reply:
I have duplicated the code as you suggested. But I don't like this. Now, if we will change something (for example will decide to save/restore bytes in the middle), we will need to change the code in three places instead of two.